File: app/src/main/java/de/leuchtetgruen/pluslocation/businessobjects/POI.java
package de.leuchtetgruen.pluslocation.businessobjects;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;

import de.leuchtetgruen.pluslocation.businessobjects.openlocationcode.CodeArea;
import de.leuchtetgruen.pluslocation.businessobjects.openlocationcode.OpenLocationCode;

@Entity(tableName = "poi", indices = {@Index("poi_code"), @Index("poi_name")})
public class POI {

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    private int id;

    @ColumnInfo(name = "poi_code")
    private String code;

    @ColumnInfo(name = "poi_name")
    private String name;
    @ColumnInfo(name = "nearby")
    private boolean usableForNearby;

    @Ignore
    private WGS84Coordinates calculatedCoordinate;

    @Ignore
    public POI(String code, String name, boolean usableForNearby) {
        this.code = code;
        this.name = name;
        this.usableForNearby = usableForNearby;
    }


    public POI(String code, String name) {
        this(code, name, false);
    }

    public WGS84Coordinates coordinate() {
        if (calculatedCoordinate == null) {
            CodeArea area = new OpenLocationCode(code).decode();
            calculatedCoordinate = new WGS84Coordinates(area.getCenterLatitude(), area.getCenterLongitude());
        }

        return calculatedCoordinate;
    }

    public int getId() {
        return id;
    }

    public String getCode() {
        return code;
    }

    public String getName() {
        return name;
    }

    public boolean isUsableForNearby() {
        return usableForNearby;
    }

    public void setId(int id) {
        this.id = id;
    }

    public void setCode(String code) {
        this.code = code;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setUsableForNearby(boolean usableForNearby) {
        this.usableForNearby = usableForNearby;
    }

    public String describe(WGS84Coordinates currentCoordinate) {
        return name + "(" + String.format("%.2f km", currentCoordinate.distanceInMeters(coordinate()) / 1000) + " " + currentCoordinate.direction(coordinate()).toString() +")";
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof  POI)) return false;

        return (((POI) obj).getCode().equals(code) &&
                ((POI) obj).getName().equals(name)
        );
    }
}

File: app/src/main/java/de/leuchtetgruen/pluslocation/businessobjects/POI.kt
package de.leuchtetgruen.pluslocation.businessobjects

import android.arch.persistence.room.*
import de.leuchtetgruen.pluslocation.businessobjects.openlocationcode.OpenLocationCode


/**
 * A POI is a Point of Interest. POIs serve two purposes. A search feature allows users to find a POI
 * that they are interested in. The app then allows them to navigate to that POI.
 *
 * POIs marked as "usable for nearby" are POIs that are commonly known (such as big train stations, landmarks etc.)
 * When a user chooses a destination the app will show nearby POIs to give the user a rough idea where
 * the chosen destination is. (E.g. think you have the PlusCode of a hotel and the app tells you it's close to the
 * main station and the zoo)
 */
@Entity(tableName = "poi", indices = arrayOf(Index("poi_code"), Index("poi_name")))
class POI @Ignore
constructor(@field:ColumnInfo(name = "poi_code")
            var code: String, @field:ColumnInfo(name = "poi_name")
            var name: String, @field:ColumnInfo(name = "nearby")
            var isUsableForNearby: Boolean) {

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    var id: Int = 0

    @Ignore
    private var calculatedCoordinate: WGS84Coordinates? = null


    constructor(code: String, name: String) : this(code, name, false)

    fun coordinate(): WGS84Coordinates {
        if (calculatedCoordinate == null) {
            val area = OpenLocationCode(code).decode()
            calculatedCoordinate = WGS84Coordinates(area.centerLatitude, area.centerLongitude)
        }

        return calculatedCoordinate!!
    }

    override fun equals(other: Any?): Boolean {
        return if (other !is POI) false else other.code == code && other.name == name

    }

    override fun hashCode(): Int {
        return "$code$name".hashCode()
    }
}
